### PR TITLE
Add video_access test

### PR DESCRIPTION
**Pull request by XianyuanLiu**
Fixes #105.

### Description
Add test_video_access.py to increase related coverage (video_access: 97%, video_datasets: 86%, videos: 47%). I have not uploaded the files causing it may not run online. Need some help. Please see the comments.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.

### CODE REVIEW

**Comment by XianyuanLiu** [comment]
I have two problems now.
1. Separate `get_train_test` will repeat the line 50-65 to get `source` and `target`. Thus, I put three tests in one function. Is it okay? Any better way?
2. I haven't uploaded the video list files. Total storage is 776k. Is it too big?

File: tests/loaddata/test_video_access.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+set_seed(seed)
+
+
+@pytest.fixture(scope="module")
+def testing_cfg():
+    cfg = CN()
+    cfg.DATASET = CN()
+    cfg.DATASET.ROOT = "E:/out/python/pt/pykale/tests/test_data/video_data/"
+    cfg.DATASET.IMAGE_MODALITY = "joint"
+    cfg.DATASET.FRAMES_PER_SEGMENT = 16
```
Comment (haipinglu) [inline tests/loaddata/test_video_access.py:34]: Is 16 the minimally possible (for the file size)? Will 8 or 4 work but lead to smaller file size?

**Comment by haipinglu** [comment]
> 1. Separate `get_train_test` will repeat the line 50-65 to get `source` and `target`. Thus, I put three tests in one function. Is it okay? Any better way?

I think it is good and I am not aware of a better way. Such good practices have been discussed in https://krokotsch.eu/cleancode/2020/08/11/Unit-Tests-for-Deep-Learning.html (search "nearly duplicates" to read that paragraph about the drawbacks of duplicates).

> 2. I haven't uploaded the video list files. Total storage is 776k. Is it too big?

You just mentioned the total. How about the max size of the individual files? I guess it should be smaller than 300k and can pass the pre-commit checks, right?

There are two options for the test data
- Direct upload: 776k seems fine. The current gait data is 277k. This is <3x. It will be good that this will be reused in other tests in future. For uploaded data, they should be reused in many to worth the fixed presence.
- Host externally and download during tests: this is what Raivo did with the video loader example, line 17-35 in https://github.com/pykale/pykale/blob/main/examples/video_loading/main.py. Actually, this code could be highly reusable and @RaivoKoot or you could consider refactoring it into `kale.loaddata` or `kale.utils` so that reuse will just need to supply the ID and filename. 

Eventually, it should be a balance between repo size and the convenience of coding (any other factor?).

I am meeting @bobturneruk in the afternoon so I will discuss with him to get some professional advice as well.

**Comment by haipinglu** [comment]
> Eventually, it should be a balance between repo size and the convenience of coding (any other factor?).
> 
> I am meeting @bobturneruk in the afternoon so I will discuss with him to get some professional advice as well.

I created a new discussion https://github.com/pykale/pykale/discussions/128 on this. @bobturneruk Let's talk about this in the afternoon.

**Comment by XianyuanLiu** [comment]
> You just mentioned the total. How about the max size of the individual files? I guess it should be smaller than 300k and can pass the pre-commit checks, right?

It is only the list files of training and test, not the video data. 776k is the total storage of the original files (8 files, each <277k). I plan to reduce the original files. It will be smaller. 

For the video data, you are right. I will change `FRAMES_PER_SEGMENT` to 4 and we only need to upload 4 frames for each dataset (not sure if it works). Each frame is 5-15k. 16 frames are 80-240k. The size is okay.

For storage. I plan to choose option 2 and add the data directory to `.gitignore`, before we have a better and direct way.

**Comment by haipinglu** [comment]
> For storage. I plan to choose option 2 and add the data directory to `.gitignore`, before we have a better and direct way.

Let's use https://github.com/pykale/data

**Comment by XianyuanLiu** [comment]
Can we directly load data from the data repo? We still need to download data to `test.test_data` and add the data directory to `.gitignore`, don't we?

**Comment by bobturneruk** [comment]
Unless you put the data in a Python package, I guess.

**Comment by RaivoKoot** [comment]
> * Actually, this code could be highly reusable and @RaivoKoot or you could consider refactoring it into `kale.loaddata` or `kale.utils` so that reuse will just need to supply the ID and filename.

Very much worth a thought. When I have some time in the near future I can research this a little bit and see if I can make it happen.

**Comment by XianyuanLiu** [comment]
Ready to be reviewed.

File: docs/source/kale.utils.rst [hunk 2]
```diff
@@ -14,6 +14,14 @@ kale.utils.csv\_logger module
    :undoc-members:
    :show-inheritance:
 
+kale.utils.download module
```
Comment (haipinglu) [inline docs/source/kale.utils.rst:17]: Why duplicate these `download` updates in two PRs and requested two reviews? This is inefficient and against our 3R principle. Try to think about the efficiency of your workflow before commits.

File: kale/utils/download.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    output_directory = Path(output_directory).absolute()
+    file = Path(output_directory).joinpath(output_file_name)
+    if os.path.exists(file):
+        logging.info("Skipping Download and Extraction")
+        return
+    if not os.path.exists(output_directory):
+        os.makedirs(output_directory)
+    logging.info("Downloading {}.".format(output_file_name))
+    download_url_to_file(url, file)
+    logging.info("Datasets downloaded in {}".format(file))
```
Comment (haipinglu) [inline kale/utils/download.py:71]: See the comments in the other PR. Confusing PRs and duplications.

**Review by haipinglu — COMMENTED** [review]
`test_video_access.py` OK but not sure how you are going to manage `download` update via duplications in two PRs. It is quite confusing to have two PRs like this (and both requested reviews), leading to inefficiency and wasted efforts.

File: kale/utils/download.py (hunk 3 above)
Comment (XianyuanLiu) [reply kale/utils/download.py:71]: > See the comments in the other PR. Confusing PRs and duplications.

Oh, my fault. I wanted to keep `add_download` separately for easy tracking. I planed to merge #132 first, and then this one. This problem will not exist. I was too confident and didn’t estimate that #132 needs some modifications last night. :)

I will fix #132 first and then merge this one.

**Review by haipinglu — APPROVED** [review]

**Pull request merged**